add npm-links sample

JiaLiPassion opened the pull request:
Add `npm-links` sample to describe how to use `esbuild` rule with `npm_link_package` rule. The sample is based on @jbedard 's sample https://github.com/jbedard/rules_esbuild/tree/npm-link-package/e2e/npm-links

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/aspect-build/rules_esbuild?pullRequest=35) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/aspect-build/rules_esbuild?pullRequest=35) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/aspect-build/rules_esbuild?pullRequest=35) it.</sub>

jbedard commented on e2e/npm-links/WORKSPACE:
```diff
@@ -0,0 +1,36 @@
+    path = "../..",
+)
+
+#---SNIP--- Below here is re-used in the workspace snippet published on releases
```
Lets drop this. Also drop a lot of the comments which are for the snippet put on the release page. Probably just copy the one from my branch.

JiaLiPassion replied:
Got it!

jbedard commented on e2e/npm-links/consume/BUILD.bazel:
```diff
@@ -0,0 +1,31 @@
+    deps = [":consumer"],
+)
+
+# Run esbuild on the consuming code
```
This runs it on the spec, not the consumer

jbedard commented on e2e/npm-links/consume/src/spec.js:
```diff
@@ -0,0 +1,4 @@
+import './consumer'
+import { ANSWER } from '@test-example/lib'
```
The spec doesn't depend on the lib, so if this was typescript being compiled it would fail. I think just drop the use of the lib from the specs?

We were trying to do esbuild -> spec -> (relative import) consumer -> (named import) lib. I don't think import is necessary to do that?

jbedard commented on e2e/npm-links/lib/BUILD.bazel:
```diff
@@ -0,0 +1,24 @@
+        ":_package_json",
+    ],
+    package = "@test-example/lib",
+    root_paths = [
```
Do we need this logic for the test? I don't think using `root_paths` should have any effect on esbuild so I don't think it's necessary?

JiaLiPassion replied:
It is just a copy from the `angular-ngc` when I want to reproduce the issue we discussed earlier. This has nothing to do with `esbuild`, and you are right, this may make the user confuse, we may add such use case example in the `npm_package` rule repo.

alexeagle commented on e2e/npm-links/WORKSPACE:
```diff
@@ -0,0 +1,36 @@
+# Override http_archive for local testing
```
at least as of the current state of the repo, anything you expect users to read and benefit from should be under `examples/` and doesn't need to be a nested workspace. You don't declare anything different in this WORKSPACE file so it's just maintenance overhead.

(there's some discussion of changing it but we haven't committed to it, and you should follow the current repo layout)

JiaLiPassion replied:
@alexeagle , there is an issue which is if we put the example to the `examples` folder and use the `WORKSPACE` file under the root, there is an error when using `npm_link_package`.

``` 
src may only be specified when linking in the root package ''
```

And I don't want to touch the `WORKSPACE` under root, not  sure is there any other way to avoid this issue.

alexeagle replied:
hmm I see, and you don't want this to use pnpm workspaces so that the lib and consume packages are registered there?

jbedard commented:
Want to also add a 3rd-party npm module that also gets imported by the `consumer`?

JiaLiPassion commented on e2e/npm-links/consume/BUILD.bazel:
```diff
@@ -0,0 +1,31 @@
+    deps = ["//:node_modules/lib"],
+)
+
+# Source consuming that named library
```
You mean this part
```
js_library(
    name = "spec",
    srcs = ["src/spec.js"],
    deps = [":consumer"],
)
```

should depends on `["src/consumer.js"]` directly?
In my understanding, `js_library` is just like `file_group`, so using `js_library` means the `spec` deps on the `:consumer` lib srcs. Is that correct?

alexeagle commented on e2e/npm-links/BUILD.bazel:
```diff
@@ -0,0 +1,18 @@
+
+LIBRARIES = [
+    "lib",
+    "consume",
```
this doesn't match the "consumer" folder, which indicates you're missing CI - if you add an e2e folder then you need to touch the .github/workflows folder, search "e2e" in rules_js to see where it goes.

JiaLiPassion replied:
Got it, here the `consumer` doesn't report error, since `consumer` doesn't have a `npm_package` target, I will add this sample to the github workflows CI.

JiaLiPassion commented:
@jbedard , @alexeagle , I updated the PR and add the sample to `ci`. 
Since the `transitive` dependency issue will not be resolved in the near future, I removed the sample about that part. 
Please review, thank you.

jbedard commented on e2e/npm-links/consumer/BUILD.bazel:
```diff
@@ -0,0 +1,54 @@
+)
+
+# Source consuming that named library
+js_library(
```
I see nothing consuming this. Does the test use this at all?

JiaLiPassion replied:
You are right, I added another rule to test the `spec bundle`.

JiaLiPassion commented on e2e/npm-links/consumer/BUILD.bazel:
```diff
@@ -0,0 +1,68 @@
+    ],
+)
+
+# Source consuming that named library
```
Yeah, you are right.
I updated the `spec.js` to add `import from lib` to test indirect import named library. Also add comments about the purpose of the test.

jbedard reviewed (approved):
make sure you squash the commits
